smac_planner/src/a_star.cpp

```diff
@@ -483,6 +645,41 @@ unsigned int & AStarAlgorithm<NodeT>::getSizeDim3()
   return _dim3_size;
 }
 
+template<typename NodeT>
+typename AStarAlgorithm<NodeT>::NodePtr AStarAlgorithm<NodeT>::tryAnalyticExpansion(
+  const NodePtr & current_node, const NodeGetter & getter, int & analytic_iterations,
+  int & closest_distance)
+{
+  if (_motion_model == MotionModel::DUBIN || _motion_model == MotionModel::REEDS_SHEPP) {
+    // This must be a NodeSE2 node if we are using these motion models
+
+    // Then check if it is collision free.
+    // Always run the expansion on the first run in case there is a
+    // trivial path to be found
+    if (analytic_iterations <= 0) {
+      // Reset the counter, and try the analytic path expansion
+      analytic_iterations = closest_distance;
+      return getAnalyticPath(current_node, getter);
+    }
+    analytic_iterations--;
+    // See if we are closer and should be expanding more often
+    const Coordinates node_coords =
+      NodeT::getCoords(current_node->getIndex(), getSizeX(), getSizeDim3());
+    closest_distance =
+      std::min(
+      closest_distance,
+      static_cast<int>(NodeT::getHeuristicCost(
+        node_coords,
+        _goal_coordinates) / NodeT::neutral_cost));
+    analytic_iterations =
```

Comment:
This all looks good to me. Can you comment on the number of expansions for a typical path in your application and how you came up with the 2.0 number? Is that a good balance between retrying often but not too often to increase the compute time a bunch?

Reply:
Another question would be if we should threshold this. Usually I'd say yes, there should be some minimum set so that we don't do every single iteration very close to the goal, but maybe that would be good. 

The 2 sides of the coin I see:
- what its doing, getting an analytic path, means that if we try close to the goal, we're pretty likely to succeed. So having it run alot close is probably moot because its the first couple tries will get something valid
- Except if you goal is near a wall and there's a room behind it, and then there will be ALOT that are close (if using the Dubin / Reeds-shepp heuristic) but are all impossible so alot of extra load. But we have the wavefront heuristic which should help with that now so maybe not a concern?

Reply:
The intuition behind a ratio of 2.0 comes from considering the case where the ratio is 1.0. In this scenario if we are 100 units from the goal we will perform the expansion in 100 iterations time. If the tree is expanding towards the goal at a rate of 1 unit distance per iteration we would reach the goal through the regular search before trying an analytic expansion. This assumes that our heuristics and costs are guiding the tree expansion to the optimal path, but with the Dubins/Reeds-Shepp heuristic and the wavefront heuristic this is close to being true. So a value greater than 1.0 is needed to guarantee that the expansion will run - if the analytic expansion is supposed to "short circuit" the tree expansion we need it to run before the tree would reach the goal anyway.

Should the default number be 2.0 or 3.0 or 4.0..? 2.0 seems like a reasonable balance where the expansion is definitely being run, but not so often that we are wasting cycles. When we are close to the goal we are running often anyway, but if the ratio is high then we run very often at the start when we are far from the goal. If we don't succeed early on, it seems likely that there are obstacles that are preventing success and the tree most likely needs to expand a bit before we have a better chance to find an analytic solution. That's as good a reason as I can give for 2.0 vs anything else. It will have a different optimal value for each scenario the planner is used in, but 2.0 seems like a default that works

Reply:
Minimum threshold is an interesting question. I fall on the side of not having a minimum because it is the analytic expansion that is allowing the paths to get to the correct goal position and avoid the issues with getting the perfect combination of quantised distances and angles to get to the goal. I think it is to our advantage to make sure the expansion is being run very often close to the goal - I think this will actually save cycles.

I can see the argument about being physically close to the goal but with an intervening obstacle, but as you point out the wavefront heuristic should all but eliminate that scenario.

Reply:
Makes sense. I can get behind 2.0 then, I was hoping for a more data-driven approach, but I agree `1 < N < 3`. 2 is a fine place to start.

On the minimum, what if the hueristic is less than the ratio, then the minimum distance would resolve to `0` and run every iteration from that point on. Which I think is a really rare edge case since that close to the goal the likelihood of problems is near-zero. I do think though just for safety thresholding to `1` or something (or hey, the `analytic_expansion_ratio` since its in the low-integers) so that we don't end up with some real problem of always running.

Reply:
Thresholding I think is the last topic to figure out before merging this :smile:

Reply:
Please run `ament_cpplint` and `colcon test --packages-select smac_planner` to make sure linting is OK (I know there's an XML error, you can ignore that) and that the new tests I added still pass. I think this change will make the `path.size()` tests different in length so feel free to update them with the current values.

Reply:
Highest frequency of expansion thresholded to a minimum 2 (every other iteration)